<a href="https://colab.research.google.com/github/Rithi-byte/PPE-DETECTION/blob/main/Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():
  print(f'User uploaded file "{fn}" with length {len(uploaded[fn])} bytes')

Saving train.zip to train.zip
User uploaded file "train.zip" with length 24420609 bytes


In [3]:
import zipfile
import os

# Get the name of the uploaded zip file (assuming only one was uploaded and it's 'train.zip')
zip_filename = 'train.zip'

# Create a ZipFile object and extract all contents
with zipfile.ZipFile(zip_filename, 'r') as zip_ref:
    zip_ref.extractall('.') # Extract to the current directory

print(f"'{zip_filename}' has been unzipped.")

# List the extracted files to show what was unzipped (optional)
print("Contents after unzipping:")
for item in os.listdir('.'):
    print(item)

'train.zip' has been unzipped.
Contents after unzipping:
.config
train
train.zip
sample_data


# Task
Explore the contents of the "train" directory, including subdirectories, to understand the dataset structure.

## Explore Training Data

### Subtask:
Inspect the contents of the `train` directory (extracted from `train.zip`) to understand the structure of the dataset. This includes identifying image files, annotation formats (if any), and available PPE categories to ensure the data is suitable for object detection.


**Reasoning**:
To understand the dataset's structure, I will use `os.walk` to traverse the `train` directory and its subdirectories, printing the contents of each level.



In [4]:
for dirpath, dirnames, filenames in os.walk('train'):
    print(f'Current Directory: {dirpath}')
    if dirnames:
        print(f'  Subdirectories: {dirnames}')
    if filenames:
        print(f'  Files: {filenames}')
    print()

Current Directory: train
  Subdirectories: ['images', 'labels']

Current Directory: train/images
  Files: ['00225_jpg.rf.e8373c3ded637a026b4ce0f51a2f181f.jpg', '00274_jpg.rf.0e26dd949e8bb93b097d7dba302461fc.jpg', 'Aitin3205_jpg.rf.3928239086b29168053e8ab386193f52.jpg', '00125_jpg.rf.07c8dd851dfef10c05cd61514a585ba3.jpg', 'Tan0827827_jpg.rf.fa4e4368247443e9714a96d67e7a05fc.jpg', 'ppe_0218_jpg.rf.7557e839d9c985dbe2b79430b69b6348.jpg', 'image_27_jpg.rf.1e9d95b619f6685a642c85af0b32cfe0.jpg', 'AightDuo0294_jpg.rf.172b37718e7a8ac828e135fb961825ec.jpg', 'ppe_0961_jpg.rf.98c4a8fe10718fb9a2844cd64e03c6f2.jpg', 'ppe_0771_jpg.rf.df0ff3556a570dc1a6e2f8e809c2f83e.jpg', '20220721_161927_jpg.rf.51d17739be30a9a5a0c097e74f37a960.jpg', 'Tan0156_jpg.rf.2d3b38cb73b7aa64c2b36390202e7382.jpg', 'construction-worker1_grande_jpg.rf.fa62d0465d2179cab25ea12d37c0482c.jpg', 'KakaoTalk_20221018_001312059_02_jpg.rf.1caa16cccf2ea0973ef231f23336a107.jpg', 'ppe_0274_jpg.rf.9c765f6413e15d0d46e7c149be8826c8.jpg', 'image_

**Reasoning**:
The previous output showed that the `train` directory contains `images` and `labels` subdirectories, with `.jpg` files in `images` and corresponding `.txt` files in `labels`. To understand the annotation format and identified PPE categories, the next step is to examine the content of one of these `.txt` label files. I will read and display the content of one sample label file to infer the annotation format and PPE categories.



In [5]:
label_file_path = 'train/labels/00225_jpg.rf.e8373c3ded637a026b4ce0f51a2f181f.txt'

print(f"Content of '{label_file_path}':")
with open(label_file_path, 'r') as f:
    print(f.read())


Content of 'train/labels/00225_jpg.rf.e8373c3ded637a026b4ce0f51a2f181f.txt':
7 0.48828125 0.565625 0.7875 0.575


The inspection of the `train` directory revealed the following structure:

- The `train` directory contains two subdirectories: `images` and `labels`.
- The `images` subdirectory contains image files (e.g., `.jpg`).
- The `labels` subdirectory contains corresponding annotation files (e.g., `.txt`).

Examination of a sample label file (`train/labels/00225_jpg.rf.e8373c3ded637a026b4ce0f51a2f181f.txt`) showed the content: `7 0.48828125 0.565625 0.7875 0.575`.

This format is consistent with the YOLO object detection annotation format, where each line represents an object with:
1.  **Class ID**: The first number (`7`) represents the category of the detected object.
2.  **Normalized Bounding Box Coordinates**: The subsequent four numbers (`0.48828125 0.565625 0.7875 0.575`) represent the `x_center`, `y_center`, `width`, and `height` of the bounding box, all normalized to the image's dimensions (values between 0 and 1).

To fully understand the dataset and identify the *available PPE categories*, a mapping from these numerical `Class IDs` to human-readable class names (e.g., `7` -> 'Hardhat') is required. This information is typically found in a `classes.txt` or `data.yaml` file. However, such a file was not found within the extracted `train` directory or its parent directory during the exploration. Without this mapping, the specific PPE categories represented by the class IDs cannot be determined.

## Install Necessary Libraries

### Subtask:
Install required Python libraries for object detection (`ultralytics` for YOLOv8) and text-to-speech conversion (`gTTS`).


**Reasoning**:
To complete the subtask of installing the required libraries, I will use `pip install` commands to install `ultralytics` and `gTTS` in a single code block.



In [ ]:
pip install ultralytics gTTS
print("Libraries 'ultralytics' and 'gTTS' installed successfully.")

SyntaxError: invalid syntax (ipython-input-35980888.py, line 1)

**Reasoning**:
The previous attempt to install libraries failed because `pip install` was not prefixed with `!` to execute it as a shell command in a Colab environment. I will add the `!` prefix to correctly install `ultralytics` and `gTTS`.



In [6]:
get_ipython().system('pip install ultralytics gTTS')
print("Libraries 'ultralytics' and 'gTTS' installed successfully.")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 28.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 kB 5.1 MB/s eta 0:00:00
  Attempting uninstall: click
    Found existing installation: click 8.3.1
    Uninstalling click-8.3.1:
      Successfully uninstalled click-8.3.1
Libraries 'ultralytics' and 'gTTS' installed successfully.


## Prepare Dataset for YOLOv8 Training

### Subtask:
Convert the inspected dataset into a format compatible with YOLOv8. This involves defining class names, creating a data.yaml configuration file, and splitting the dataset into training and validation sets.


**Reasoning**:
The first step in preparing the dataset for YOLOv8 training is to define the class names and create the `data.yaml` configuration file. This file is crucial for YOLOv8 to understand the dataset structure and the classes it needs to detect. I will use the assumed class names and specify the paths for the training and validation data that will be created in subsequent steps.



In [7]:
import yaml

# Define the assumed class names based on common PPE items.
# The class ID '7' observed in the label file will be mapped to 'goggles' for this example.
class_names = [
    'hardhat',   # 0
    'vest',      # 1
    'gloves',    # 2
    'boots',     # 3
    'person',    # 4
    'earmuffs',  # 5
    'mask',      # 6
    'goggles'    # 7 (based on the sample label file)
]

# Create the data.yaml content
data_yaml_content = {
    'path': '/content/', # Dataset root directory
    'train': 'train_split/images', # Train images relative to 'path'
    'val': 'val_split/images',   # Val images relative to 'path'
    'names': class_names
}

# Define the path for data.yaml
data_yaml_path = 'data.yaml'

# Write the content to data.yaml
with open(data_yaml_path, 'w') as f:
    yaml.dump(data_yaml_content, f, default_flow_style=False)

print(f"'{data_yaml_path}' created successfully with content:")
print(yaml.dump(data_yaml_content, default_flow_style=False))


'data.yaml' created successfully with content:
names:
- hardhat
- vest
- gloves
- boots
- person
- earmuffs
- mask
- goggles
path: /content/
train: train_split/images
val: val_split/images



**Reasoning**:
Now that the `data.yaml` file has been created, the next step in preparing the dataset for YOLOv8 training is to split the dataset into training and validation sets. This involves creating the necessary directory structure for the splits and then moving the image and label files from the original `train` directory into these new directories, maintaining a consistent split ratio.



In [8]:
import os
import shutil
import random

# Define paths
original_images_dir = 'train/images'
original_labels_dir = 'train/labels'

train_images_dir = 'train_split/images'
train_labels_dir = 'train_split/labels'
val_images_dir = 'val_split/images'
val_labels_dir = 'val_split/labels'

# Create target directories if they don't exist
for d in [train_images_dir, train_labels_dir, val_images_dir, val_labels_dir]:
    os.makedirs(d, exist_ok=True)

# Get all image files
image_files = [f for f in os.listdir(original_images_dir) if f.lower().endswith(('.png', '.jpg', '.jpeg'))]
random.shuffle(image_files)

# Define split ratio (e.g., 80% train, 20% validation)
split_ratio = 0.8
split_index = int(len(image_files) * split_ratio)

train_files = image_files[:split_index]
val_files = image_files[split_index:]

print(f"Total images: {len(image_files)}")
print(f"Training images: {len(train_files)}")
print(f"Validation images: {len(val_files)}")

# Function to move files
def move_files(file_list, source_image_dir, source_label_dir, dest_image_dir, dest_label_dir):
    for filename in file_list:
        # Move image
        shutil.copy(os.path.join(source_image_dir, filename), os.path.join(dest_image_dir, filename))

        # Move corresponding label (assuming same base name, .txt extension)
        label_filename = os.path.splitext(filename)[0] + '.txt'
        if os.path.exists(os.path.join(source_label_dir, label_filename)):
            shutil.copy(os.path.join(source_label_dir, label_filename), os.path.join(dest_label_dir, label_filename))
        else:
            print(f"Warning: Label file {label_filename} not found for image {filename}")

# Move files to train and validation directories
print("Moving training files...")
move_files(train_files, original_images_dir, original_labels_dir, train_images_dir, train_labels_dir)

print("Moving validation files...")
move_files(val_files, original_images_dir, original_labels_dir, val_images_dir, val_labels_dir)

print("Dataset split into training and validation sets successfully.")

Total images: 368
Training images: 294
Validation images: 74
Moving training files...
Moving validation files...
Dataset split into training and validation sets successfully.


## Train YOLOv8 Model for PPE Detection

### Subtask:
Initialize a pre-trained YOLOv8 object detection model and fine-tune it on your prepared PPE dataset. This will involve setting training parameters like the number of epochs, batch size, and learning rate to achieve optimal performance for detecting gloves, vests, boots, goggles, and helmets.


**Reasoning**:
To initialize and train the YOLOv8 model, I will import the `YOLO` class from `ultralytics` and load the `yolov8n.pt` pre-trained model. Then, I will start the training process by calling the `.train()` method, passing the `data.yaml` configuration, the number of epochs, and the image size as specified in the instructions.



In [9]:
from ultralytics import YOLO

# Load a pre-trained YOLOv8n model
model = YOLO('yolov8n.pt')  # You can choose other models like yolov8s.pt, yolov8m.pt, etc.

# Train the model
# Set the number of epochs, image size, and data configuration file.
results = model.train(data='data.yaml', epochs=50, imgsz=640)

print("YOLOv8 model training initiated successfully.")

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
Ultralytics 8.4.14 🚀 Python-3.12.12 torch-2.9.0+cpu CPU (Intel Xeon CPU @ 2.20GHz)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=data.yaml, degrees=0.0, deterministic=True, device=cpu, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, ke